duckduckgo/content-scope-scripts
# ntp: virtual list for favorites

**Asana Task/Github Issue:** https://app.asana.com/0/1201141132935289/1208792034336660/f

## Description

- [x] only render the rows of favorites that are visible
  - this addresses a major pain-point from our native desktop browsers
- [x] when expanded, defer rendering children until the main thread is free
  - this means users on the very slowest devices still get an instant feedback on click 

## Example
Part of this PR is optimizing the expand/collapse action of the favorites list. On slow devices this can cause over a second of input delay since the act of rendering a screen full of favorites can be an expensive main-thread-blocking operation.

To combat this problem, I separated the expand/collapse action from the rendering of the actual tiles, which had the following impact: 💪🏻 

**BEFORE** (20x CPU slowdown)
<img width="647" alt="Screenshot 2024-11-26 at 10 13 02 AM" src="https://github.com/user-attachments/assets/5ac5d3f5-0653-441f-a2b9-9781e9902597">

**AFTER** (20x CPU slowdown)

<img width="540" alt="Screenshot 2024-11-26 at 10 22 32 AM" src="https://github.com/user-attachments/assets/b2bc679c-f9c8-4342-abb6-1c551ba79190">



## TODO:
- [x] add some perf metrics to explain the change

## Testing Steps

- https://deploy-preview-1269--content-scope-scripts.netlify.app/build/pages/new-tab/?favorites=100
- Verify that the page works as expected with showing/hiding along with drag+drop

## Checklist


*Please tick all that apply:*
